satsense/features/texton.py

```diff
@@ -58,23 +59,39 @@ def get_texton_descriptors(image: Image):
 Image.register('texton_descriptors', get_texton_descriptors)
 
 
-def texton_cluster(images: Iterator[Image], n_clusters=32,
-                   sample_size=100000) -> MiniBatchKMeans:
+def texton_cluster(images: Iterator[Image],
+                   n_clusters=32,
+                   sample_size=100000,
+                   sample_window=(8192, 8192)) -> MiniBatchKMeans:
     """Compute texton clusters."""
+    nfeatures = int(sample_size / len(images))
     descriptors = []
     for image in images:
-        array = image['texton_descriptors']
-        array = array.reshape(-1, array.shape[-1])
-        non_masked = ~array.mask.any(axis=-1)
-        descriptors.append(array.data[non_masked])
+        if image.shape[0] < sample_window[0]:
+            sample_window = (image.shape[0], sample_window[1])
+        if image.shape[1] < sample_window[1]:
+            sample_window = (sample_window[0], image.shape[1])
+
+        generator = FullGenerator(image, sample_window)
+        generator.load_image('texton_descriptors', (sample_window, ))
+
+        img_descriptors = []
+        for img in generator:
```

Comment:
```suggestion
        for array in generator:
```
and remove the next line